File: py/openaxis/tests/test_logging.py
import datetime as dt
import logging
from pathlib import Path
import subprocess
import sys

import pytest

from openaxis.logging import DiagnosticLog, configure_logging


def test_session_header_identifies_sdk_client_and_survives_rotation(tmp_path):
    from openaxis._version import __version__
    log = DiagnosticLog('header-test', directory=tmp_path, client_version='2.3.4', max_bytes=1)
    header = log.path.read_text()
    assert f'OpenAxis SDK {__version__} (Python); client=header-test; client_version=2.3.4' in header
    log.info('first')
    log.info('second')
    assert log.path.read_text().startswith(header)
    assert Path(str(log.path) + '.1').read_text().startswith(header)
    assert log.path.read_text().endswith('INFO second\n')
    log.close()


def test_collision_rotation_and_mirror(tmp_path, monkeypatch):
    from openaxis import logging as module
    class Frozen(dt.datetime):
        @classmethod
        def now(cls, tz=None): return cls(2026, 9, 19, 14, 25, 30, tzinfo=tz)
    monkeypatch.setattr(module.dt, "datetime", Frozen)
    seen = []
    a = DiagnosticLog("fusion", directory=tmp_path, max_bytes=80, sinks=[lambda *v: seen.append(v)])
    b = DiagnosticLog("fusion", directory=tmp_path)
    try:
        assert a.path.name == "fusion-20260919T142530Z.log"
        assert b.path.name == "fusion-20260919T142530Z-2.log"
        a.info("first " + "x" * 40)
        a.info("second " + "y" * 40)
        assert "first" in Path(str(a.path) + ".1").read_text()
        assert "second" in a.path.read_text()
        assert "OpenAxis SDK" in b.path.read_text()
        assert len(seen) == 2
    finally:
        a.close(); b.close()


def test_cleanup_keeps_ten_sessions_without_active_protection(tmp_path):
    active = DiagnosticLog("fusion", directory=tmp_path)
    active.info("still running")
    created = {active.path}
    for _ in range(12):
        log = DiagnosticLog("fusion", directory=tmp_path)
        assert log.path not in created
        created.add(log.path)
        log.close()
    try:
        assert not active.path.exists()
        assert len(list(tmp_path.glob("*.log"))) == 10
        assert not list(tmp_path.glob("*.lock"))
        mirrored = []
        active.sinks.append(lambda *record: mirrored.append(record))
        active.info("after cleanup")
        assert active.error and mirrored == [("info", "after cleanup")]
        assert not active.path.exists()
    finally:
        active.close()


def test_other_process_can_clean_old_session(tmp_path):
    active = DiagnosticLog("fusion", directory=tmp_path)
    try:
        subprocess.run([sys.executable, "-c",
            "from openaxis.logging import DiagnosticLog; import sys; "
            "s=DiagnosticLog('fusion',directory=sys.argv[1],keep=0); s.close()", str(tmp_path)], check=True)
        assert not active.path.exists()
    finally:
        active.close()
    cleanup = DiagnosticLog("fusion", directory=tmp_path, keep=0)
    try:
        assert not active.path.exists()
    finally:
        cleanup.close()


def test_sdk_events_route_automatically(tmp_path):
    sink = configure_logging("test-app", directory=tmp_path)
    from openaxis.navigation_diagnostics import NavigationDiagnostics
    logging.getLogger("openaxis").warning("transport error")
    NavigationDiagnostics().observe("gesture_started", gesture_id=1)
    text = sink.path.read_text()
    assert "transport error" in text and "gesture_started" in text
    sink.close()


def test_disk_failure_still_mirrors(tmp_path):
    path = tmp_path / "not-a-directory"
    path.write_text("existing data")
    seen = []
    sink = DiagnosticLog("test", directory=path, sinks=[lambda *v: seen.append(v)])
    sink.info("message")
    assert sink.error and seen == [("info", "message")]
    sink.close()


@pytest.mark.parametrize("name", ["../escape", "UPPER", "", "a/b"])
def test_invalid_client(name, tmp_path):
    with pytest.raises(ValueError): DiagnosticLog(name, directory=tmp_path)


def test_shared_logging_contract(tmp_path):
    import json, re
    from openaxis.logging import format_record, normalize_level
    fixture = json.loads((Path(__file__).parents[3] / "fixtures/openaxis-1.0/logging.json").read_text())
    now = dt.datetime.fromtimestamp(fixture["epoch_ms"] / 1000, dt.timezone(dt.timedelta(minutes=fixture["offset_minutes"])))
    records = []
    sink = DiagnosticLog("conformance", directory=tmp_path, sinks=[lambda *v: records.append(v)])
    expected = ""
    try:
        for record in fixture["records"]:
            assert normalize_level(record["level"]) == record["normalized"]
            assert format_record(record["level"], record["message"], now) == record["line"]
            sink.write(record["level"], record["message"])
            if record["normalized"] in fixture["debug_off"]: expected += record["line"]
        complete = sink.path.read_bytes().decode("utf-8")
        actual = complete.split("\n", 1)[1]
        assert re.sub(r"\d{4}-\d\d-\d\d \d\d:\d\d:\d\d\.\d{3} [+-]\d\d:\d\d", fixture["stamp"], actual) == expected
        before = len(records)
        sink.close(); sink.write("info", "after close")
        assert len(records) == before
        assert sink.path.read_bytes().decode("utf-8") == complete
        for name in fixture["invalid_clients"]:
            with pytest.raises(ValueError): DiagnosticLog(name, directory=tmp_path)
        configured = [configure_logging(name, directory=tmp_path) for name in fixture["configuration"]]
        assert configured[0] is configured[2] and configured[0] is not configured[1]
        logging.getLogger("openaxis").info("reactivated")
        assert "reactivated" in configured[0].path.read_text()
        assert "reactivated" not in configured[1].path.read_text()
        configured[0].close()
        fresh = configure_logging(fixture["configuration"][0], directory=tmp_path)
        assert fresh is not configured[0]
        fresh.close(); configured[1].close()
    finally:
        sink.close()


def test_shared_native_retention_and_boundary(tmp_path):
    import json
    f = json.loads((Path(__file__).parents[3] / "fixtures/openaxis-1.0/logging.json").read_text())
    for i in range(1, 13):
        name = "retention-20200101T000000Z" + ("" if i == 1 else f"-{i}") + ".log"
        (tmp_path / name).write_bytes(b"")
    log = DiagnosticLog("retention", directory=tmp_path)
    assert log.max_bytes == f["defaults"]["max_bytes"]
    expected = {f"retention-20200101T000000Z-{i}.log" for i in f["retained_suffixes"]}
    assert {p.name for p in tmp_path.glob("retention-2020*.log")} == expected
    log.close()
    r = f["rotation"]
    log = DiagnosticLog("boundary", directory=tmp_path, max_bytes=r["max_bytes"])
    header_bytes = log.path.stat().st_size
    log.max_bytes += header_bytes
    for _ in range(r["writes_before_rotation"]): log.info(r["message"])
    assert log.path.stat().st_size == r["max_bytes"] + header_bytes
    assert not Path(str(log.path) + ".1").exists()
    log.info(r["message"])
    assert Path(str(log.path) + ".1").stat().st_size == r["max_bytes"] + header_bytes
    log.close()

File: cpp/tests/test_logging.cpp
#include <openaxis/logging.hpp>
#include <cassert>
#include <iostream>
#include <nlohmann/json.hpp>

int main(int argc, char **argv) {
    if (argc == 4 && std::string(argv[1]) == "--log-probe") {
        openaxis::DiagnosticLog log("interop", argv[2], 100, std::string(argv[3]) == "cleanup" ? 0 : 5);
        if (log.path().empty()) return 2;
        log.write("info", "probe");
        std::cout << log.path().string() << std::endl;
        if (std::string(argv[3]) == "hold") { std::string line; std::getline(std::cin, line); }
        return 0;
    }
    namespace fs = std::filesystem;
    auto root = fs::temp_directory_path() / ("openaxis-log-test-" + std::to_string(
        std::chrono::steady_clock::now().time_since_epoch().count()));
    {
        using Log = openaxis::DiagnosticLog;
        std::ifstream cmakeFile(std::string(OPENAXIS_FIXTURES) + "/../../cpp/CMakeLists.txt");
        std::string cmake((std::istreambuf_iterator<char>(cmakeFile)), {});
        std::smatch version;
        assert(std::regex_search(cmake, version, std::regex("project\\(OpenAxis VERSION ([0-9.]+)")));
        nlohmann::json f; std::ifstream(std::string(OPENAXIS_FIXTURES) + "/logging.json") >> f;
        auto now = std::chrono::system_clock::time_point(std::chrono::milliseconds(f["epoch_ms"].get<long long>()));
        fs::create_directories(root);
        for (int i = 1; i <= 12; ++i) {
            auto name = "retention-20200101T000000Z" + (i == 1 ? std::string() : "-" + std::to_string(i)) + ".log";
            std::ofstream(root / name).put(' ');
        }
        { Log retention("retention", root);
          std::set<std::string> expected, actual;
          for (auto &i : f["retained_suffixes"]) expected.insert("retention-20200101T000000Z-" + std::to_string(i.get<int>()) + ".log");
          for (auto &e : fs::directory_iterator(root)) if (e.path().extension() == ".log" && e.path().filename().string().find("retention-2020") == 0) actual.insert(e.path().filename().string());
          assert(expected == actual);
        }
        { auto r = f["rotation"];
          auto header_bytes = Log::format_record("info", "OpenAxis SDK 1.0.0 (C++); client=boundary; client_version=unknown").size();
          Log boundary("boundary", root, r["max_bytes"].get<std::size_t>() + header_bytes);
          for (int i=0; i<r["writes_before_rotation"].get<int>(); ++i) boundary.write("info", r["message"].get<std::string>());
          assert(fs::file_size(boundary.path()) == r["max_bytes"].get<std::size_t>() + header_bytes);
          assert(!fs::exists(boundary.path().string() + ".1"));
          boundary.write("info", r["message"].get<std::string>());
          assert(fs::file_size(boundary.path().string() + ".1") == r["max_bytes"].get<std::size_t>() + header_bytes);
        }
        Log log("conformance", root);
        { std::ifstream file(log.path()); std::string header; std::getline(file, header);
          assert(header.find("OpenAxis SDK " + version[1].str() + " (C++)") != std::string::npos); }
        int mirrors = 0; log.sink = [&](auto &, auto &) { ++mirrors; };
        std::string expected;
        for (const auto &record : f["records"]) {
            auto level = record["level"].get<std::string>(), message = record["message"].get<std::string>();
            assert(Log::normalize_level(level) == record["normalized"]);
            assert(Log::format_record(level, message, now, f["offset_minutes"].get<int>()) == record["line"]);
            log.write(level, message);
            if (record["normalized"] != "debug") expected += record["line"].get<std::string>();
        }
        std::ifstream input(log.path(), std::ios::binary);
        std::string actual((std::istreambuf_iterator<char>(input)), {}); input.close();
        actual = actual.substr(actual.find('\n') + 1);
        assert(std::regex_replace(actual, std::regex("[0-9]{4}-[0-9]{2}-[0-9]{2} [0-9]{2}:[0-9]{2}:[0-9]{2}\\.[0-9]{3} [+-][0-9]{2}:[0-9]{2}"), f["stamp"].get<std::string>()) == expected);
        auto count = mirrors; log.close(); log.write("info", "after close"); assert(mirrors == count);
        for (auto &client : f["invalid_clients"]) {
            bool rejected = false;
            try { Log invalid(client.get<std::string>(), root); } catch (const std::invalid_argument &) { rejected = true; }
            assert(rejected);
        }
        std::vector<std::shared_ptr<Log>> configured;
        for (auto &client : f["configuration"]) configured.push_back(Log::configure(client.get<std::string>(), root));
        assert(configured[0] == configured[2] && configured[0] != configured[1]);
        configured[0]->close(); auto fresh = Log::configure("alpha", root); assert(fresh != configured[0]);
        fresh->close(); configured[1]->close();
    }
    {
        openaxis::DiagnosticLog active("test", root, 100);
        assert(!active.path().empty());
        int mirrors = 0;
        active.sink = [&](auto &, auto &) { ++mirrors; };
        active.write("info", std::string(80, 'x'));
        active.write("warning", "new record");
        assert(fs::exists(active.path().string() + ".1"));
        assert(mirrors == 2);
        {
            openaxis::DiagnosticLog other("test", root);
            assert(active.path() != other.path());
        }
        for (int i = 0; i < 12; ++i) { openaxis::DiagnosticLog old("test", root); }
        assert(!fs::exists(active.path()));
        int count = 0;
        for (auto &entry : fs::directory_iterator(root)) if (entry.path().extension() == ".log" && entry.path().filename().string().find("test-") == 0) ++count;
        assert(count == 10);
    }
    {
        openaxis::DiagnosticLog clean("test", root, 100, 0);
        int count = 0;
        for (auto &entry : fs::directory_iterator(root)) if (entry.path().extension() == ".log" && entry.path().filename().string().find("test-") == 0) ++count;
        assert(count == 1);
    }
    {
        openaxis::DiagnosticLog log("header-test", root, 1, 5, "2.3.4");
        log.write("info", "first"); log.write("info", "second");
        for (const auto &path : {log.path(), fs::path(log.path().string() + ".1")}) {
            std::ifstream input(path); std::string header; std::getline(input, header);
            assert(header.find("OpenAxis SDK 1.0.0 (C++); client=header-test; client_version=2.3.4") != std::string::npos);
        }
    }
    fs::remove_all(root);
    std::cout << "Diagnostic log tests passed\n";
}
